[java] Update rule AvoidArrayLoops

adangel opened the pull request:
## Describe the PR

- Part of #2701

pmd-test commented:
<!--
  0 Errors
  0 Warnings
  2 Messages: Compared to pmd/7.0.x:<br />
T...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          2 Messages
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">Compared to pmd/7.0.x:<br />
This changeset changes 0 violations,<br />
introduces 134 new violations, 0 new errors and 0 new configuration errors,<br />
removes 0 violations, 0 errors and 0 configuration errors.<br />
<a href="https://chunk.io/pmd/1e6e3a47c3c74b5d8291799e86dc407e/diff1/index.html">Full report</a></td>
    </tr>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">Compared to master:<br />
This changeset changes 30227 violations,<br />
introduces 32803 new violations, 3 new errors and 0 new configuration errors,<br />
removes 143848 violations, 8 errors and 3 configuration errors.<br />
<a href="https://chunk.io/pmd/1e6e3a47c3c74b5d8291799e86dc407e/diff2/index.html">Full report</a></td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

adangel commented:
1. false positive:

```java
table[i] = labels[offset + readInt(u)];
```

(https://github.com/spring-projects/spring-framework/blob/v5.0.6.RELEASE/spring-core/src/main/java/org/springframework/asm/ClassReader.java#L1595)

That's not simply done by calling System.arraycopy. The array index on the rhs is calculated via readInt... and it is not based on i (so, no fixed offset)

2. false positive

```java
S[i] = S[j];
S[j] = Si;
```
(https://github.com/openjdk/jdk/blob/jdk-11%2B28/src/java.base/share/classes/com/sun/crypto/provider/ARCFOURCipher.java#L78)

Not so simple either. It's the same array - so the content is shifted. Not a use case for Arrays.copyOf...

3. valid, correct case

```java
                for (int i=0; i<time.length; i++) {
                    blob[8+time.length-i-1] = time[i];
                }
```

(https://github.com/openjdk/jdk/blob/jdk-11%2B28/src/java.base/share/classes/com/sun/security/ntlm/Client.java#L168)

I think, that's actually correct. This could indeed be rewritten as: `System.arraycopy(time, 0, blob, 8+time.length-1, time.length)`.

4. false positive

```
        eachArg:
            for (int j = 0; j < arguments.length; j++) {
                if (arguments[j] instanceof Name) {
                    Name n = (Name) arguments[j];
                    int check = n.index;
                    // harmless check to see if the thing is already in newNames:
                    if (check >= 0 && check < newNames.length && n == newNames[check])
                        continue eachArg;
                    // n might not have the correct index: n != oldNames[n.index].
                    for (int i = start; i < end; i++) {
                        if (n == oldNames[i]) {
                            if (n == newNames[i])
                                continue eachArg;
                            if (!replaced) {
                                replaced = true;
                                arguments = arguments.clone();
                            }
                            arguments[j] = newNames[i];
                            continue eachArg;
                        }
                    }
                }
            }
```

(https://github.com/openjdk/jdk/blob/jdk-11%2B28/src/java.base/share/classes/java/lang/invoke/LambdaForm.java#L1453)

There are two for-loops. I guess, we detected the inner loop, but for some reason, we reported the outer loop. The copy might be `arguments[j] = newNames[i];`, but a `continue` statement is following, which makes this not a straightforward copy...

5. false positive

```java
        for (int i = 0; i < uVlen; i++) {
            int nextp = a[headp];
            a[headp] = uValues[i];
            headp = nextp;
        }
```

(https://github.com/openjdk/jdk/blob/jdk-11%2B28/src/java.base/share/classes/com/sun/java/util/jar/pack/PopulationCoding.java#L299)

for the array assignment on the lhs the loop index is not used.

6. false positive

```java
        for (int i = srcBegin + (1 >> LO_BYTE_SHIFT); i < srcEnd; i += 2) {
            dst[dstBegin++] = value[i];
        }
```

(https://github.com/openjdk/jdk/blob/jdk-11%2B28/src/java.base/share/classes/java/lang/StringUTF16.java#L263)
